### ISSUE TITLE
Calls to coroutines when guarded by inspect.iscoroutinefunction are reported as errors

### ISSUE THREAD

**Issue by IAmTomahawkx** [microsoft/pyright#3423]
**Describe the bug**
Type guarding using inspect.iscoroutinefunction (or asyncio.iscoroutinefunction) does not affect calls to the coroutine after the guard

**To Reproduce**
See code below

**Expected behavior**
Ideally this shouldn't result in a type error

**Screenshots or Code**
```python
import inspect
from typing import List, Union, Coroutine, Any

T = Union[List[str], Coroutine[Any, Any, None]]

async def foo(x: T) -> Any:
    if inspect.iscoroutinefunction(x):
        await x()  # Object of type "List[str]" is not callable
```
<img width="587" alt="image" src="https://user-images.githubusercontent.com/47765953/166617229-a532d846-24c7-49cc-9122-135c342d8544.png">

**VS Code extension or command-line**
VSC extension `v2022.4.3`

**Comment by JelleZijlstra** [comment]
This should be fixed by using a TypeGuard in typeshed: https://github.com/python/typeshed/blob/master/stdlib/inspect.pyi#L186. Feel free to open an issue or even a PR over on the typeshed repo.
